rclcpp_action/include/rclcpp_action/client_goal_handle_impl.hpp

```diff
@@ -168,7 +168,7 @@ ClientGoalHandle<ActionT>::call_feedback_callback(
     RCLCPP_ERROR(rclcpp::get_logger("rclcpp_action"), "Sent feedback to wrong goal handle.");
     return;
   }
-  std::lock_guard<std::mutex> guard(handle_mutex_);
```

Comment:
Alternative idea, do we need to hold the lock here? What if we hold the lock in a scope `{...}` only long enough to copy `feedback_callback_` to a local variable, then release the lock before calling the callback? I think that solves the issue and avoids (probably small) overhead of using a recursive mutex.

Reply:
I'm not sure that will work.

There are a couple of assumptions at work here:
1.  That the `shared_this` being passed in here is directly the same as `this`.  That is a weird API, but OK.  In that case, we can consider that everything that would need to be protected from concurrent access to `this` also applies to `shared_this`.
2.  That the `handle_mutex_` here is protecting all members of this object from concurrent modification/access.

Taken together, I think it means that we need to hold the mutex across the entire operation (since `feedback_callback_` is passed `shared_this`).  Does that make sense?  Or am I maybe missing something?

Reply:
> That the shared_this being passed in here is directly the same as this. That is a weird API, but OK. In that case, we can consider that everything that would need to be protected from concurrent access to this also applies to shared_this.

If `shared_this` isn't `this` then holding a lock on `this->handle_mutex_` isn't going to help as the feedback callback is given `shared_this`, not `this`. I think we can ignore this though because there is a check that `shared_this.get() == this` above.

> That the `handle_mutex_` here is protecting all members of this object from concurrent modification/access.
> [...] Taken together, I think it means that we need to hold the mutex across the entire operation [...]

IIUC holding the lock during the feedback's entire call would protect against a situation where the goal handle changed state between two feedback calls. Just getting an idea for what could change, it looks like all the public methods are:

```
  /// Get the unique ID for the goal.
  const GoalUUID &
  get_goal_id() const;

  /// Get the time when the goal was accepted.
  rclcpp::Time
  get_goal_stamp() const;

  /// Get the goal status code.
  int8_t
  get_status();

  /// Check if an action client has subscribed to feedback for the goal.
  bool
  is_feedback_aware();

  /// Check if an action client has requested the result for the goal.
  bool
  is_result_aware()
```

All of those methods only "read" data from the class, so the concern would be a "write" somewhere else at the same time as the feedback callback. "writes" could come from the `Client` [as it's a friend of this class](https://github.com/ros2/rclcpp/blob/c6edffafefa477068f3e00d4e65b83a720195eef/rclcpp_action/include/rclcpp_action/client_goal_handle.hpp#L107).

I'm choosing to ignore `get_goal_id()` and `get_goal_stamp()` as they're not even protected by locks. They assume `info_` is constant (side note: [I think this could be enforced by adding const here](https://github.com/ros2/rclcpp/blob/c6edffafefa477068f3e00d4e65b83a720195eef/rclcpp_action/include/rclcpp_action/client_goal_handle.hpp#L154)).

`get_status` [is protected by a lock](https://github.com/ros2/rclcpp/blob/c6edffafefa477068f3e00d4e65b83a720195eef/rclcpp_action/include/rclcpp_action/client_goal_handle_impl.hpp#L101). If we wrap the `feedback_callback_` in the lock then it's possible the feedback callback could call get status twice and get different results (executing one call, then complete the next).

`is_feedback_aware()` [is protected by a lock](https://github.com/ros2/rclcpp/blob/c6edffafefa477068f3e00d4e65b83a720195eef/rclcpp_action/include/rclcpp_action/client_goal_handle_impl.hpp#L117) and means "does this goal handle have a feedback callback"? If the callback is being executed, it already knows the answer and doesn't have a need to call it. I suppose it's possible the feedback callback could get set to a `nullptr` while this one is running and `is_feedback_aware()` might return `false` if this function called it, but I wasn't able to find any calls to `client_goal_handle->set_feedback_callback()`  anywhere (side note: if it's private and no one calls it, can we delete it?).

`is_result_aware()` [is protected by a lock](https://github.com/ros2/rclcpp/blob/c6edffafefa477068f3e00d4e65b83a720195eef/rclcpp_action/include/rclcpp_action/client_goal_handle_impl.hpp#L125) and means "does someone want the result of this goal"? It's possible  that someone could write a feedback callback that they gave access to the `Client` and only asked for the result if the feedback reports a certain value. If they were using  multithreaded executor and put the Client into a Reentrant callback group and feedback messages came in fast enough then two feedback callbacks could be executed simultaneously and both would call [client->async_get_result()](https://github.com/ros2/rclcpp/blob/c6edffafefa477068f3e00d4e65b83a720195eef/rclcpp_action/include/rclcpp_action/client.hpp#L495C3-L495C19). There can only be one result callback, [so the second's feedback callback would replace the first's](https://github.com/ros2/rclcpp/blob/c6edffafefa477068f3e00d4e65b83a720195eef/rclcpp_action/include/rclcpp_action/client.hpp#L508-L511).

Affecting decision making in the `feedback_callback` based on `is_result_aware()` seems like the only potential negative consequence to me, and it requires the user passing in the `Client` to the callback some other way.

All that to say, I'm fine with the current approach, but I don't see any significant problems with releasing the lock during the feedback callback either. All holding the lock seems to do is make sure two feedback callbacks on one goal handle can't be run in parallel regardless of the executor or callback group.

Reply:
> side note: if it's private and no one calls it, can we delete it?

I was thinking the same thing, i guess we can just remove this. but originally this should have been exposed to user application to change the feedback callback? i was not sure about that.

Reply:
I think releasing the mutex is a bad idea.
If a reentrant cb group is used, not holding the mutex could lead to strange stuff like the handle->getStatus() function changing the result in the middle of the feedback callback.
I guess this is the case, why the call_feedback_callback function exists in the first place.

Reply:
@sloretz Thanks for the detailed analysis, that is very helpful.  Here are a few comments:

> IIUC holding the lock during the feedback's entire call would protect against a situation where the goal handle changed state between two feedback calls.

To be clear, we are already holding the lock across the `feedback_callback_` right now.  The difference in this PR is that we are switching that to a `recursive_mutex`.  But your overall point stands that we could consider dropping the mutex before making that call.

Considering your points, I agree that given the code today we could probably drop the lock without too many negative consequences.  But I think I still prefer to keep holding the lock for two reasons:

1.  @jmachowinski point about things changing while the callback is running.  We've run into these kinds of problems before in `rclcpp`, particularly around parameters, so we also hold a lock around that for similar reasons.
2.  In the future it is possible someone could add a method that did a "write", and it would be hard to figure out that it breaks our current assumptions.

So all of that is to say that I'm good with this PR as-is.  I'm going to do a review here, and then we can run CI on it.